docs/kb/wasm.rst

```diff
@@ -0,0 +1,307 @@
+
+Scylla accepts UDF's source code in WebAssembly text format - also known as ``WAT``. The source can use and define whatever's needed for execution, including multiple helper functions and symbols. The requirements for it to be accepted as correct UDF source is that the WebAssembly module exports a symbol with the same name as the function, this symbol's type is indeed a function with correct signature, the module exports a ``_scylla_abi`` global and all symbols related to the selected ABI version.
+
+UDF's source code can be, naturally, simply coded by hand in ``WAT``. It is not often very convenient to program directly in assembly, so here are a few tips.
+
+Compiling to Wasm
+^^^^^^^^^^^^^^^^^
+
+Rust
+....
+
+The main supported language for Wasm UDFs is Rust. To generate WebAssembly from rust, it's best to use the `scylla-udf <https://github.com/scylladb/scylla-rust-udf>`_ Rust helper library, and follow the instructions present there.
+
+As a short example, here's a sample Rust code which can be compiled to WebAssembly:
+
+.. code-block:: rust
+
+    use scylla_udf::export_udf;
+
+    #[export_udf]
+    pub fn fib(n: i32) -> i32 {
+        if n < 2 {
+            n
+        } else {
+            fib(n - 1) + fib(n - 2)
+        }
+    }
+
+The compilation instructions are described at https://github.com/scylladb/scylla-rust-udf but the commands will generally be:
+
+.. code-block:: bash
+
+    cargo build --target=wasm32-wasi
+    wasm2wat target/wasm32-wasi/debug/fib.wasm > fib.wat
+
+
+C
+.
+
+Clang is capable of compiling C source code to Wasm and it also supports useful built-ins
+for using Wasm-specific interfaces, like ``__builtin_wasm_memory_size`` and ``__builtin_wasm_memory_grow``
+for memory management.
+
+However, there is no C helper library yet, so implementing UDFs in C is in general much more difficult than in Rust.
+Just to implement the fib() function, you need something like this:
+
+.. code-block:: c
+
+    #include<stdlib.h>
+
+    const int WASM_PAGE_SIZE = 64 * 1024;
+
+    const int _scylla_abi = 2;
+
+    void* _scylla_malloc(int size) {
+        return malloc(size);
+    }
+
+    void _scylla_free(void* ptr) {
+        free(ptr);
+    }
+
+    static long long swap_int64(long long val) {
+        val = ((val << 8) & 0xFF00FF00FF00FF00ULL ) | ((val >> 8) & 0x00FF00FF00FF00FFULL );
+        val = ((val << 16) & 0xFFFF0000FFFF0000ULL ) | ((val >> 16) & 0x0000FFFF0000FFFFULL );
+        return (val << 32) | ((val >> 32) & 0xFFFFFFFFULL);
+    }
+
+    long long fib_aux(long long n) {
+        if (n < 2) {
+            return n;
+        }
+        return fib_aux(n-1) + fib_aux(n-2);
+    }
+
+    long long fib(long long p) {
+        int size = p >> 32;
+        long long* p_val = (long long*)(p & 0xffffffff);
+        // Initialize memory for the return value
+        long long* ret_val = _scylla_malloc(sizeof(long long));
+        if (size == -1) {
+            *ret_val = swap_int64(42);
```

Comment:
What's 42? The meaning of life? :-) I don't understand what this case does.

Reply:
I only now see that it actually makes no sense with current description (I thought you were referring to the `swap` calls). The example was supposed to perform some special action when encountering a NULL instead of also returning NULL (we have a `RETURNS NULL ON NULL INPUT` clause for that), so we returned a special value 42 for that instead.

I think I'll use a different example for the Rust function, so that there is no confusion about

Reply:
I only now see that it actually makes no sense with current description (I thought you were referring to the swap calls). The example was supposed to perform some special action when encountering a `NULL` instead of also returning `NULL` (we have a `RETURNS NULL ON NULL INPUT` clause for that), so we returned a special value 42 for that instead.

I think I'll use a different example for the Rust function, so that there is no confusion about comparing it to the C code. It will probably also make more sense, because the current example is relatively easy also in C, but the new one will actually make use of the Rust helper macro capabilities